mountpoint-s3-client/src/s3_crt_client/put_object.rs

```diff
@@ -106,6 +126,47 @@ pub struct S3PutObjectRequest {
     review_callback: ReviewCallbackBox,
     start_time: Instant,
     total_bytes: u64,
+    // Headers of the CompleteMultipartUpload response, available after the request was finished
+    response_headers: Arc<Mutex<Option<Headers>>>,
+    // Server-side encryption type which is expected to be found in response_headers
+    pub server_side_encryption: Option<String>,
+    /// Server-side encryption KMS key ID which is expected to be found in response_headers
+    pub ssekms_key_id: Option<String>,
+}
+
+/// If non empty `server_side_encryption` or `ssekms_key_id` were used, this function checks headers
+/// of the CompleteMultipartUpload response to contain the expected values
+fn check_response_headers(
+    response_headers: Arc<Mutex<Option<Headers>>>,
```

Comment:
Especially if you end up testing this function directly, I'd change this to `Option<&Headers>` and move the locking + unwrapping up the stack (where the assumptions are clearer!).